cpp/src/io/orc/writer_impl.cu

```diff
@@ -925,10 +925,14 @@ void writer::impl::write_data_stream(gpu::StripeStream const &strm_desc,
 {
   const auto length                                        = strm_desc.stream_size;
   (*streams)[enc_stream.ids[strm_desc.stream_type]].length = length;
-  if (length != 0) {
-    const auto *stream_in = (compression_kind_ == NONE)
-                              ? enc_stream.data_ptrs[strm_desc.stream_type]
-                              : (compressed_data + strm_desc.bfr_offset);
+  if (length == 0) { return; }
+
+  const auto *stream_in = (compression_kind_ == NONE) ? enc_stream.data_ptrs[strm_desc.stream_type]
+                                                      : (compressed_data + strm_desc.bfr_offset);
+
+  if (out_sink_->is_device_write_preferred(length)) {
+    out_sink_->device_write(stream_in, length, stream);
```

Comment:
If device write is preferred then we don't need `stream_out`. I see `stream_out` is a pinned host allocation. So, add a check there and avoid it?

Reply:
You want to avoid the allocation of `stream_out` if device write is preferred?

Reply:
That is possible right?

Reply:
It's a bit messy, but possible. We could make `stream_output` an `optional` and initialize it in a function that checks if device reads are preferred for **all** streams (doesn't need to be a separate function, but `write` is large enough).

Reply:
I might not have understood @devavret's suggestion; it appears that the `cudaMemcpyAsync()` is in the `! is_device_write_preferred()` path. So this conversation can be resolved?